hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/OpcodesController.java

```diff
@@ -0,0 +1,86 @@
+import com.hedera.mirror.web3.service.ContractCallService;
+import com.hedera.mirror.web3.service.OpcodeService;
+import io.github.bucket4j.Bucket;
+import jakarta.validation.Valid;
+import lombok.CustomLog;
+import lombok.RequiredArgsConstructor;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@CustomLog
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/contracts/results")
+@ConditionalOnProperty(prefix = "hedera.mirror.opcode.tracer", name = "enabled", havingValue = "true")
+class OpcodesController {
+
+    private final OpcodeService opcodeService;
+    private final ContractCallService contractCallService;
+    private final Bucket rateLimitBucket;
+    private final Bucket gasLimitBucket;
+
+    /**
+     * <p>
+     *     Returns a result containing detailed information for the transaction execution,
+     *     including all values from the {@code stack}, {@code memory} and {@code storage}
+     *     and the entire trace of opcodes that were executed during the replay.
+     * </p>
+     * <p>
+     *     Note that to provide the output, the transaction needs to be re-executed on the EVM,
+     *     which may take a significant amount of time to complete if stack and memory information is requested.
+     * </p>
+     *
+     * @param transactionIdOrHash The transaction ID or hash
+     * @param stack Include stack information
+     * @param memory Include memory information
+     * @param storage Include storage information
+     * @return {@link OpcodesResponse} containing the result of the transaction execution
+     */
+    @GetMapping(value = "/{transactionIdOrHash}/opcodes")
+    OpcodesResponse getContractOpcodes(
+            @PathVariable @Valid TransactionIdOrHashParameter transactionIdOrHash,
+            @RequestParam(required = false, defaultValue = "true") boolean stack,
+            @RequestParam(required = false, defaultValue = "false") boolean memory,
+            @RequestParam(required = false, defaultValue = "false") boolean storage
+    ) {
+        if (!rateLimitBucket.tryConsume(1)) {
+            throw new RateLimitException("Rate limit exceeded.");
+        }
+
+        final var params = opcodeService.buildCallServiceParameters(transactionIdOrHash);
+        if (!gasLimitBucket.tryConsume(params.getGas())) {
+            throw new RateLimitException("Rate limit exceeded.");
+        }
+
+        final var options = new OpcodeTracerOptions(stack, memory, storage);
+        final var result = contractCallService.processOpcodeCall(params, options);
+
+        return opcodeService.buildOpcodesResponse(result);
```

Comment:
I think it's awkward that the opcode service only builds request and response and doesn't have responsibility to invoke the call service. Think we should encapsulate opcode flow entirely within opcode service.

```suggestion
        final var options = new OpcodeTracerOptions(stack, memory, storage);
        return opcodeService.opcodeCall(transactionIdOrHash, options);
```

Reply:
Also, it's debatable but I think the mapping to view model response should be moved back to the controller.

Reply:
I don't have a strong opinion on this one I just moved it to the service because I thought it would be better to also cover it with the service integration tests because we are calling the database accessor service and we only have unit tests for the controller. But if you prefer, I can move this back to the controller.